feat(Insomnia): custom linting rules for cloud and local (#5287)
Kong/developer.konghq.com: app/_how-tos/insomnia/add-custom-linting-rules.md

---
title: "Add custom linting rules in {{ site.data.products.insomnia.name }}"
permalink: /how-to/add-custom-linting-rules/

content_type: how_to
description: Learn how to add custom linting rules to your APIs in Insomnia .
no_wrap: true

products:
  - insomnia 

min_version:
  insomnia: "13.0"
tags:
  - insomnia-documents
  - linting

related_resources:
  - text: "Design APIs with {{ site.data.products.insomnia.name }}"
    url: /insomnia/design/

tldr:
  q: How do I customize linting in {{ site.data.products.insomnia.name }} ?
  a: In your {{ site.data.products.insomnia.name }}  document, click the upload icon for the **Default OAS Ruleset** and add a Spectral ruleset YAML file with your custom ruleset.

prereqs:
  inline:
  - title: Create a design document
    include_content: prereqs/design-document
    icon_url: /assets/icons/file.svg

faqs:
  - q: How can I use custom linting with Inso CLI?
    a: |
      Create a [Spectral ruleset](https://docs.stoplight.io/docs/spectral/e5b9616d6d50c-rulesets) in YAML format, and upload it to an OAS file document in Insomnia by clicking the upload icon for the **Default OAS Ruleset**.

---

> {{ site.data.products.insomnia.name }}  provides a default linting ruleset. Override it to add your custom linting rules by following these steps:

Reviewer: ```suggestion
{{ site.data.products.insomnia.name }}  provides a default linting ruleset. Override it to add your custom linting rules by following these steps.
```

---
title: "Add custom linting rules in {{ site.data.products.insomnia.name }}"
permalink: /how-to/add-custom-linting-rules/

content_type: how_to

> description: Learn how to add custom linting rules to your APIs in {{ site.data.products.insomnia.name }} .

Reviewer: ```suggestion
description: Learn how to add custom linting rules to your APIs in Insomnia .
```
I think we can leave this one as-is since I don't think description is liquified?

no_wrap: true

products:
  - insomnia 

min_version:
  insomnia: "13.0"
tags:
  - insomnia-documents
  - linting

related_resources:
  - text: "Design APIs with {{ site.data.products.insomnia.name }}"
    url: /insomnia/design/

tldr:
  q: How do I customize linting in {{ site.data.products.insomnia.name }} ?

>   a: In your {{ site.data.products.insomnia.name }}  project, add a YAML file containing your custom ruleset.

Reviewer: ```suggestion
  a: In your {{ site.data.products.insomnia.name }}  document, click the upload icon for the **Default OAS Ruleset** and add a Spectral ruleset YAML file with your custom ruleset.
```

prereqs:
  inline:
  - title: Create a design document
    include_content: prereqs/design-document
    icon_url: /assets/icons/file.svg

faqs:
  - q: How can I use custom linting with Inso CLI?
    a: |

>       Create a [Spectral ruleset](https://docs.stoplight.io/docs/spectral/e5b9616d6d50c-rulesets) in YAML format, and upload it in the same directory as the OAS file to lint.

Reviewer: ```suggestion
      Create a [Spectral ruleset](https://docs.stoplight.io/docs/spectral/e5b9616d6d50c-rulesets) in YAML format, and upload it to an OAS file document in Insomnia by clicking the upload icon for the **Default OAS Ruleset**.
```

---

{{ site.data.products.insomnia.name }}  provides a default linting ruleset. Override it to add your custom linting rules by following these steps:

## Upload the ruleset file

From the {{ site.data.products.insomnia.name }}  app, upload a [Spectral ruleset](https://docs.stoplight.io/docs/spectral/e5b9616d6d50c-rulesets) in YAML format. Upload it from the project containing the design document with the OpenAPI specifications (OAS) to lint.

This places the ruleset file in the local working directory. {{ site.data.products.insomnia.name }}  renames this custom ruleset as `.spectral.yaml`.

> You can view the content of the file by clicking `Custom Ruleset` in {{ site.data.products.insomnia.name }} , but not edit the file from the UI.

Reviewer: ```suggestion
1. In the {{ site.data.products.insomnia.name }}  app, navigate to a document with the OAS file you want to lint.
2. Click the upload icon for **Default OAS Ruleset** at the bottom of the pane.

This places the ruleset file in the local working directory. {{ site.data.products.insomnia.name }}  renames this custom ruleset as `.spectral.yaml`.

You can view the content of the file by clicking `Custom Ruleset` in {{ site.data.products.insomnia.name }}. If you want to make changes to the file, edit it locally.
```
The last sentence is a guess but trying to emulate Lena and frame things in a positive way to tell them how they can edit the file.

Author: resolved in b7cc455bee6619f70a3fa8df9020f28ef07fe0a3

> ## Define the rules

Reviewer: I can't leave a suggestion since some lines are unedited, but let's move this whole "Define the rules" section so it goes before the "Upload the ruleset file" section. That way users know what to put in the file and then they can upload it.

Author: resolved in 328a851f2

tldr: 
  q: How do I customize linting in Insomnia?

>   a: In your Insomnia project, add a YAML file containing your custom ruleset.

Reviewer 2: (as of 13, you can do this in the UI)

Reviewer 2: also worth noting, perhaps as an FAQ, that as of 13 you can lint in any project type, not just git/local.

Author: Updated

>   a: In your Insomnia project, add a YAML file containing your custom ruleset.

Reviewer 3: Should we be more explicit and say: "containing your custom **Spectral** ruleset"

Author: resolved with `ee7d90764`

prereqs:
  inline:
  - title: Create a design document
    include_content: prereqs/design-document
    icon_url: /assets/icons/file.svg

> faqs:

Reviewer 3: Should we include a FAQ about backwards compatibility? Consider two users on a cloud sync project. User A is on v13, User B is on v12.6. User A uploads a custom ruleset and syncs it to project. User B attempts to pull and they get an error because their build does not know what custom rulesets are. In that case User B will need to update to v13. @chayglass84 I still need to verify exactly what happens in this scenario (it is part of my testing plan) but this is my hypothesis.

Author: added warning

  - q: How can I use custom linting with Inso CLI?
    a: |

>       Create your YAML file in the same directory as the OAS file to lint, then run the [`inso lint spec`](/inso-cli/reference/lint_spec/) command.

Reviewer 3: In v13 users will be able to define a `--ruleset` flag for Inso. When this flag is passed, it will override any `.spectral.yaml` files in the root directory. 

`-r, --ruleset <path>  path to a Spectral ruleset file, overrides default OAS ruleset and any ruleset in the API Spec folder`

If the flag is not specified, the ruleset defined in `.spectral.yaml` in the root is used (if it exists). Otherwise, the default OAS ruleset will be applied.

Author: Resolved in new version

---

## Upload the ruleset file

Upload a YAML file containing your ruleset to the project containing the design document with the OpenAPI specifications (OAS) to lint.

> This places the ruleset file in the local working directory. You don't see the file in the Insomnia UI, but the linting rules are applied to the associated OAS.

Reviewer 3: The content of the selected ruleset file can be viewed in a modal via the UI by clicking on the `Custom Ruleset` button. Note the modal is read-only and the user cannot edit ruleset from there. 

<img width="1131" height="1310" alt="Image" src="https://github.com/user-attachments/assets/171c1dfa-ea3f-4861-b147-0738b6ab5165" />

Author: resolved in `ccc385b50`

## Define the rules

The custom ruleset overrides the default one. To create a completely new ruleset, add your rules in the file using the [Spectral](https://docs.stoplight.io/docs/spectral/e5b9616d6d50c-rulesets) syntax. If you want to extend an existing ruleset, specify the ruleset with the `extends` property.

{:.info}

> > Available Spectral properties in Insomnia are `rules` and `extends`.

Reviewer 3: Is it possible to be more detailed here? Suggestion below. 

Available top level Spectral properties in Insomnia include `rules` and `extends`. Custom functions are not permitted; however Insomnia provides support for [Spectral built-in core functions](https://docs.stoplight.io/docs/spectral/cb95cf0d26b83-core-functions).

Author: resolved in `6ef9b3c8f`

      function: truthy
```

## Validate

> Close and reopen the document to apply the changes. In this example, you can confirm the rule fires by adding a tag without a description:

Reviewer: ```suggestion
In this example, you can confirm the rule triggers by removing the description for the `GET /flights` endpoint. This will be flagged since the ruleset doesn't allow tags without a description:
```
It auto applied for me after I uploaded it.

```yaml
tags:

>   - name: flight-data

Reviewer: ```suggestion
"/flights":
    get:
      summary: Get KongAir planned flights
      tags:
      - flight-data
```
Changing the example since the other one didn't work for me

```

> This causes a new warning to appear:

Reviewer: Removing the specific language because mine had different verbage.
```suggestion
This displays an error stating that tags must have a description.
```

Author: Yes, because you used my custom example, but the one in the docs sets it to a warning, not error.

```
tag-description Tags must have a description.
```

> ## Override the rules

Author: ```suggestion
## Override the rules {% new_in 13 %} 
```

If the `--ruleset` flag isn't specified, {{ site.data.products.insomnia.name }}  uses one of the following, in order:

- The ruleset defined in `.spectral.yaml`, if it exists.
- The default OAS ruleset.

{% endnavtab %}
{% navtab "Use `extends` in `.spectral.yaml`" %}

Make {{ site.data.products.insomnia.name }}  point at another ruleset in `.spectral.yaml` by using the `extends` property. For example:

```yaml
extends:
  - spectral:oas
  - ./rules/my-rules.yaml
```

If you reference a remote file (one hosted outside of your project in {{ site.data.products.insomnia.name }}), {{ site.data.products.insomnia.name }} fetches its content only when you upload the custom ruleset. If the remote file later changes, {{ site.data.products.insomnia.name }}:

- Doesn't track those changes in commits, because the file lives outside the project.
- Doesn't fetch them automatically for security reasons. Instead, you can refresh manually using the button next to the ruleset option.

{% endnavtab %}
{% endnavtabs %}

{:.warning}

> > Make sure users in your team are running at least {{ site.data.products.insomnia.name }} `13.0` to avoid compatibility issues with this feature.

Reviewer: Is it just the override that is part of 13? If so, we can add the {% new_in 13 %} badge to that heading and remove this note.

Author: This note seems important to me because some compatibility features might arise if any user isn't using v13. The person reading this page might not understand where the problems come from because they are using v13 on their machine, but this note will make them find out the origins of some problems and check that the rest of the team has upgraded the version as well.

Reviewer 4: I think we should use the new_in badge here. The doc will have: 
```
min_version:
  insomnia: "13.0"
```
Making it clear that it is only compatible with Insomnia 13, and the new_in badge makes it feel like a feature where an admonition makes it feel like a warning.
